### PR TITLE
[#1678] fix(server): disk size leak on removing resources by AppPurgeEvent

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

Descrease the disk size  that calculated by local storage self on removing resources with AppPurgeEvent

### Why are the changes needed?

Fix: #1678

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by zuston** [comment]
This is also should be cherry picked into branch-0.9.0

**Comment by zuston** [comment]
cc @rickyma

**Comment by rickyma** [comment]
If I get this correctly, this should be a bug that is easily encountered in a production environment, unless you don't primarily write to local disks in your production env.

**Comment by zuston** [comment]
> If I get this correctly, this should be a bug that is easily encountered in a production environment, unless you don't primarily write to local disks in your production env.

You can see the test case to understand this. And this has happened in our internal env.

**Comment by rickyma** [comment]
I've seen it already, that's why I think this should be a common issue. It should happen easily in a prod env. Isn't removing app happening all the time, is it?

**Comment by zuston** [comment]
> I've seen it already, that's why I think this should be a common issue. It should happen easily in a prod env. Isn't removing app happening all the time, is it?

Yes. But I think this will happen when AppPurgeEvent is triggered and ShufflePurgeEvent is not triggered at the same time, maybe this is rare.

**Comment by rickyma** [comment]
> Yes. But I think this will happen when AppPurgeEvent is triggered and ShufflePurgeEvent is not triggered at the same time, maybe this is rare.

ShufflePurgeEvent will only be created when an unregister shuffle request is received.
So I think it will help a lot in my case. This is very common in my env. Because Spark jobs might be killed by users, which will cause the threads interrupted and fail to continue to send unregister shuffle requests to shuffle servers. The stack trace is something like:
```
[21:57:36:865] [unregister-shuffle-0] WARN  org.apache.uniffle.client.impl.ShuffleWriteClientImpl.lambda$unregisterShuffle$28:1021 - Error happened when unregistering to ShuffleServerInfo{host[123.123.123.123], grpc port[12345], netty port[17000]}
io.grpc.StatusRuntimeException: CANCELLED: Thread interrupted
        at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:268) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:249) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:167) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at org.apache.uniffle.proto.ShuffleServerGrpc$ShuffleServerBlockingStub.unregisterShuffleByAppId(ShuffleServerGrpc.java:772) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at org.apache.uniffle.client.impl.grpc.ShuffleServerGrpcClient.doUnregisterShuffleByAppId(ShuffleServerGrpcClient.java:337) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at org.apache.uniffle.client.impl.grpc.ShuffleServerGrpcClient.unregisterShuffleByAppId(ShuffleServerGrpcClient.java:344) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at org.apache.uniffle.client.impl.ShuffleWriteClientImpl.lambda$unregisterShuffle$28(ShuffleWriteClientImpl.java:1016) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at org.apache.uniffle.common.util.ThreadUtils.lambda$null$0(ThreadUtils.java:110) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_352]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_352]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_352]
        at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_352]
Caused by: java.lang.InterruptedException
        at io.grpc.stub.ClientCalls$ThreadlessExecutor.throwIfInterrupted(ClientCalls.java:750) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at io.grpc.stub.ClientCalls$ThreadlessExecutor.waitAndDrain(ClientCalls.java:711) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:159) ~[rss-client-spark3-shaded-0.9.0-SNAPSHOT.jar:?]
        ... 9 more
```

**Comment by zuston** [comment]
This is a easy-to-understand bug. Could you help review this? @rickyma

File: server/src/test/java/org/apache/uniffle/server/ShuffleTaskManagerTest.java [hunk 1]
```diff
@@ -104,6 +104,67 @@ public void afterEach() throws Exception {
...
+    conf.set(ShuffleServerConf.JETTY_HTTP_PORT, 12345);
+    conf.set(ShuffleServerConf.JETTY_CORE_POOL_SIZE, 64);
+    conf.set(ShuffleServerConf.SERVER_BUFFER_CAPACITY, 1000L);
+    conf.set(ShuffleServerConf.SERVER_MEMORY_SHUFFLE_HIGHWATERMARK_PERCENTAGE, 50.0);
+    conf.set(ShuffleServerConf.SERVER_MEMORY_SHUFFLE_LOWWATERMARK_PERCENTAGE, 0.0);
+    conf.set(ShuffleServerConf.SERVER_COMMIT_TIMEOUT, 10000L);
+    conf.set(ShuffleServerConf.SERVER_APP_EXPIRED_WITHOUT_HEARTBEAT, 100000L);
+    conf.set(ShuffleServerConf.HEALTH_CHECK_ENABLE, false);
+
+    conf.setString(ShuffleServerConf.RSS_STORAGE_TYPE.key(), "LOCALFILE");
```
[the remark this answers is not in the document]
Comment (zuston) [reply server/src/test/java/org/apache/uniffle/server/ShuffleTaskManagerTest.java]: done

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java [hunk 2]
```diff
@@ -771,7 +772,9 @@ public void removeResources(String appId, boolean checkAppExpired) {
       partitionsToBlockIds.remove(appId);
       shuffleBufferManager.removeBuffer(appId);
       shuffleFlushManager.removeResources(appId);
-      storageManager.removeResources(new AppPurgeEvent(appId, shuffleTaskInfo.getUser()));
+      storageManager.removeResources(
+          new AppPurgeEvent(
```
Comment (xianjingfeng) [inline server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:776]: Should we remove the following constructor? For prevent the same problem from happening again.
https://github.com/apache/incubator-uniffle/blob/917456c5eb0d7b70f7faa5da489aaa29d54cde40/server/src/main/java/org/apache/uniffle/server/event/AppPurgeEvent.java#L28-L30

File: server/src/test/java/org/apache/uniffle/server/ShuffleTaskManagerTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rickyma) [reply server/src/test/java/org/apache/uniffle/server/ShuffleTaskManagerTest.java]: Can you refactor all the other codes in `ShuffleTaskManagerTest.java` using `constructServerConfWithLocalfile`? If you search for `String confFile = ClassLoader.getSystemResource("server.conf").getFile()`, you will find a lot of duplicated codes.

**Comment by jerqi** [comment]
If the disk is broken, the shuffle may fail to remove. Will the size be calculated correctly?

**Comment by zuston** [comment]
> If the disk is broken, the shuffle may fail to remove. Will the size be calculated correctly? I haven't seen similar test case.

This is another case and that is why I prefer not using the disk size that uniffle maintained.

File: server/src/test/java/org/apache/uniffle/server/ShuffleTaskManagerTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zuston) [reply server/src/test/java/org/apache/uniffle/server/ShuffleTaskManagerTest.java]: Let’s put this optimization in another PR as the good first issue

**Comment by xianjingfeng** [comment]
> > If the disk is broken, the shuffle may fail to remove. Will the size be calculated correctly? I haven't seen similar test case.
> 
> This is another case and that is why I prefer not using the disk size that uniffle maintained.

+1. I'm planning to add a switch to skip determine whether it is writable by using the disk size in metaData.

**Comment by xianjingfeng** [comment]
> This is another case and that is why I prefer not using the disk size that uniffle maintained.

Could we use this variable instead?  @zuston @jerqi 
https://github.com/apache/incubator-uniffle/blob/43bd09c652889919ad888d31d4e3a5461c93f21d/storage/src/main/java/org/apache/uniffle/storage/common/LocalStorage.java#L49

**Comment by xianjingfeng** [comment]
> This is another case and that is why I prefer not using the disk size that uniffle maintained.

Here we have calculated the space usage of the server, could we use it for judgment?
https://github.com/apache/incubator-uniffle/blob/43bd09c652889919ad888d31d4e3a5461c93f21d/server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java#L116

File: server/src/test/java/org/apache/uniffle/server/ShuffleTaskManagerTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rickyma) [reply server/src/test/java/org/apache/uniffle/server/ShuffleTaskManagerTest.java]: OK.

**Comment by jerqi** [comment]
> > This is another case and that is why I prefer not using the disk size that uniffle maintained.
> 
> Here we have calculated the space usage of the server, could we use it for judgment? @jerqi @zuston
> 
> https://github.com/apache/incubator-uniffle/blob/43bd09c652889919ad888d31d4e3a5461c93f21d/server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java#L116

This can't handle mix deployment with other services.

**Comment by zuston** [comment]
> > This is another case and that is why I prefer not using the disk size that uniffle maintained.
> 
> Here we have calculated the space usage of the server, could we use it for judgment? @jerqi @zuston
> 
> https://github.com/apache/incubator-uniffle/blob/43bd09c652889919ad888d31d4e3a5461c93f21d/server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java#L116

+1 for this.

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java (hunk 2 above)
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:776]: This will bring too many changes.

**Comment by xianjingfeng** [comment]
> > > This is another case and that is why I prefer not using the disk size that uniffle maintained.
> > 
> > 
> > Here we have calculated the space usage of the server, could we use it for judgment? @jerqi @zuston
> > https://github.com/apache/incubator-uniffle/blob/43bd09c652889919ad888d31d4e3a5461c93f21d/server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java#L116
> 
> This can't handle mix deployment with other services.

No, it can.
https://github.com/apache/incubator-uniffle/blob/43bd09c652889919ad888d31d4e3a5461c93f21d/server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java#L200-L224

**Comment by jerqi** [comment]
> > > > This is another case and that is why I prefer not using the disk size that uniffle maintained.
> > > 
> > > 
> > > Here we have calculated the space usage of the server, could we use it for judgment? @jerqi @zuston
> > > https://github.com/apache/incubator-uniffle/blob/43bd09c652889919ad888d31d4e3a5461c93f21d/server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java#L116
> > 
> > 
> > This can't handle mix deployment with other services.
> 
> No, it can.
> 
> https://github.com/apache/incubator-uniffle/blob/43bd09c652889919ad888d31d4e3a5461c93f21d/server/src/main/java/org/apache/uniffle/server/LocalStorageChecker.java#L200-L224

Ok for me.

**Comment by jerqi** [comment]
Another concern is performance. Is it related to system call?

**Comment by rickyma** [comment]
> Another concern is performance. Is it related to system call?

Yeah, do you know the original background of `LocalStorageMeta`? Was it designed to store disk size in `LocalStorageMeta` to maintain performance, considering that frequent access to disk size might affect performance?

**Comment by jerqi** [comment]
> > Another concern is performance. Is it related to system call?
> 
> Yeah, do you know the original background of `LocalStorageMeta`? Was it designed to store disk size in `LocalStorageMeta` to maintain performance, considering that frequent access to disk size might affect performance?

No, this is used for uploading data from the disk to HDFS in the origin design.

**Review by xianjingfeng — APPROVED** [review]
LGTM

**Review by rickyma — APPROVED** [review]

**Comment by zuston** [comment]
Merged. Thanks for your review @rickyma @jerqi @xianjingfeng

**Pull request merged**